runtime/reconcilers/model.go

```diff
@@ -1258,40 +1287,97 @@ func (r *ModelReconciler) executeSingle(ctx context.Context, executor *wrappedMo
 		if err != nil {
 			return nil, err
 		}
-		stageDuration = stageResult.ExecDuration
+		finalResult.ExecDuration += stageDuration
+
+		return finalResult, nil
+	})
+}
 
-		// We change the inputProps to be the result properties of the stage step
-		inputProps = stageResult.Properties
+// executeWithRetry applies retry logic around the provided execution function.
+func (r *ModelReconciler) executeWithRetry(ctx context.Context, self *runtimev1.Resource, mdl *runtimev1.Model, executeFunc func(context.Context) (*drivers.ModelResult, error)) (*drivers.ModelResult, error) {
+	// Apply defaults for retry options
+	var defaultAttempts uint32 = 3
+	var defaultDelay uint32 = 5
+	defaultExponentialBackoff := true
+	defaultIfErrorMatches := []string{".*OvercommitTracker.*", ".*Bad Gateway.*", ".*Timeout.*"}
 
-		// Drop the stage result after the final step has executed.
-		// We do this using the same ctx, which means we may leak data in the staging connector in case of context cancellations.
-		// This is acceptable since the staging connector is assumed to be configured for temporary data.
-		defer func() {
-			err := executor.stageResultManager.Delete(ctx, stageResult)
-			if err != nil {
-				r.C.Logger.Warn("Failed to clean up staged model output", zap.String("model", self.Meta.Name.Name), zap.Error(err), observability.ZapCtx(ctx))
+	retryAttempts := mdl.Spec.RetryAttempts
+	if retryAttempts == nil {
+		retryAttempts = &defaultAttempts
+	}
+	retryDelay := mdl.Spec.RetryDelay
+	if retryDelay == nil {
+		retryDelay = &defaultDelay
+	}
+	retryExponentialBackoff := mdl.Spec.RetryExponentialBackoff
+	if retryExponentialBackoff == nil {
+		retryExponentialBackoff = &defaultExponentialBackoff
+	}
+	retryIfErrorMatches := mdl.Spec.RetryIfErrorMatches
+	if len(retryIfErrorMatches) == 0 {
+		retryIfErrorMatches = defaultIfErrorMatches
+	}
+
+	attempts := int(*retryAttempts)
+	if attempts == 0 {
+		attempts = 1
+	}
+
+	// maxRetries is not used, so remove its assignment
+	backoff := time.Duration(*retryDelay) * time.Second
+	if backoff == 0 {
+		backoff = 5 * time.Second
+	}
```

Comment:
This default is already applied above. This case only runs if the user explicitly set it to 0 – if they did that, isn't the intention just to keep it 0, i.e. retry immediately?